File: backend-ruby/lib/elasticsearch_client.rb
require "json"
require "net/http"
require "uri"

class ElasticsearchClient
  def initialize(base_url)
    @base_url = base_url.sub(%r{/+\z}, "")
  end

  def health
    started = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    uri = URI(@base_url)
    response = request(Net::HTTP::Get.new(uri), uri, timeout: 3)
    payload = JSON.parse(response.body)
    {
      ok: true, status: "ok", backend: "ruby",
      latency_ms: elapsed_ms(started),
      cluster_name: payload["cluster_name"].to_s,
      version: payload.dig("version", "number").to_s
    }
  rescue StandardError => e
    { ok: false, status: "error", backend: "ruby", latency_ms: elapsed_ms(started), error: e.message }
  end

  def search(index:, query:, timeout: 15, page: 1, size: 20)
    uri = URI("#{@base_url}/#{index.to_s.gsub(%r{\A/+|/+\z}, "")}/_search")
    uri.query = URI.encode_www_form(ignore_unavailable: "true")
    http_request = Net::HTTP::Post.new(uri)
    http_request["Content-Type"] = "application/json"
    http_request.body = JSON.generate(
      query: query,
      sort: [{ "@timestamp" => { order: "desc", unmapped_type: "date" } }],
      from: (page - 1) * size, size: size, track_total_hits: true, timeout: "5s",
      _source: ["@timestamp", "host", "program", "msg", "severity", "dt", "hr"]
    )
    body = JSON.parse(request(http_request, uri, timeout: timeout).body)
    raise "Elasticsearch search failed: #{body["error"]}" if body["error"]

    total_value = body.dig("hits", "total") || 0
    {
      "total" => total_value.is_a?(Hash) ? total_value.fetch("value", 0).to_i : total_value.to_i,
      "hits" => body.dig("hits", "hits") || []
    }
  end

  private

  def elapsed_ms(started)
    ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - started) * 1000).round
  end

  def request(request, uri, timeout:)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == "https"
    http.open_timeout = timeout
    http.read_timeout = timeout
    response = http.request(request)
    unless response.is_a?(Net::HTTPSuccess) || response.is_a?(Net::HTTPRedirection)
      raise "Elasticsearch HTTP #{response.code}: #{response.body}"
    end
    response
  end
end

File: backend-java/src/main/java/com/example/flaskelastic/ElasticsearchClient.java
package com.example.flaskelastic;

import com.fasterxml.jackson.core.type.TypeReference;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

final class ElasticsearchClient implements QueryClient {
    private final HttpClient client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();
    private final URI baseUri;

    ElasticsearchClient(Config config) {
        this.baseUri = URI.create(config.elasticsearchUrl().replaceAll("/+$", ""));
    }

    @Override
    public Map<String, Object> info() throws Exception {
        HttpRequest request = HttpRequest.newBuilder(baseUri).timeout(Duration.ofSeconds(3)).GET().build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        if (response.statusCode() < 200 || response.statusCode() >= 400) {
            throw new IOException("elasticsearch info failed: HTTP " + response.statusCode());
        }
        return App.JSON.readValue(response.body(), new TypeReference<>() {});
    }

    @Override
    public ElasticSearchResult search(String index, Map<String, Object> query, int page, int size) throws Exception {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("query", query);
        body.put("sort", List.of(Map.of("@timestamp", Map.of("order", "desc", "unmapped_type", "date"))));
        body.put("from", (page - 1) * size);
        body.put("size", size);
        body.put("track_total_hits", true);
        body.put("timeout", "5s");
        body.put("_source", List.of("@timestamp", "host", "program", "msg", "severity", "dt", "hr"));

        URI searchUri = URI.create(baseUri + "/" + index.replaceAll("^/+|/+$", "") + "/_search?ignore_unavailable=true");
        HttpRequest request = HttpRequest.newBuilder(searchUri)
                .timeout(Duration.ofSeconds(15))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(App.JSON.writeValueAsString(body), StandardCharsets.UTF_8))
                .build();
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new IOException("elasticsearch search failed: HTTP " + response.statusCode() + ": " + response.body());
        }
        Map<String, Object> decoded = App.JSON.readValue(response.body(), new TypeReference<>() {});
        if (decoded.containsKey("error")) {
            throw new IOException("elasticsearch search failed: " + decoded.get("error"));
        }
        return decodeResult(decoded);
    }

    @SuppressWarnings("unchecked")
    private ElasticSearchResult decodeResult(Map<String, Object> decoded) {
        List<SearchHit> results = new ArrayList<>();
        if (!(decoded.get("hits") instanceof Map<?, ?> hitsMap)
                || !(hitsMap.get("hits") instanceof List<?> hitRows)) {
            return new ElasticSearchResult(0, results);
        }
        Object totalValue = hitsMap.get("total");
        int total = totalValue instanceof Map<?, ?> totalMap
                ? Integer.parseInt(Values.string(totalMap.get("value")))
                : Integer.parseInt(Values.string(totalValue));
        for (Object hitRow : hitRows) {
            Map<?, ?> hit = (Map<?, ?>) hitRow;
            Map<String, Object> source = hit.get("_source") instanceof Map<?, ?> sourceMap
                    ? new LinkedHashMap<>((Map<String, Object>) sourceMap) : new LinkedHashMap<>();
            results.add(new SearchHit(
                    Values.string(hit.get("_id")), Values.string(hit.get("_index")), source, hit.get("_score")
            ));
        }
        return new ElasticSearchResult(total, results);
    }
}
